File: libsrc/arrivals.f90
module Arrivals

use datetime_module,   only: datetime, timedelta
use hypfile

implicit none
private

type, public     :: cRecordArr
   integer          :: id = 0
   character(LEN=5) :: sta = ''
   character(LEN=5) :: phase = ''
   real             :: trec = 0.0
   real             :: wt = 0.0
   real             :: amp = 0.0
   real             :: freq = 0.0
   integer          :: key = 0
   real             :: X
   real             :: Y
   real             :: Z
contains
!procedure ::
end type cRecordArr


type, public      :: cFileArr
   integer           :: n_arr = 0
   type (datetime)   :: reftime
   real              :: sumwt = 0.0
   real              :: sumwt2 = 0.0
   real              :: maxwt = 0.0
   type(cRecordArr), allocatable,  dimension(:)  :: arr
contains
procedure :: fill => cfilearr_fill
procedure :: toDateTime => cfilearr_todatetime
end type cFileArr

type(cFileArr),pointer,public,save   :: parrs

contains

subroutine cfilearr_todatetime(this,rtime,year,month,day,hour,minute,isec,msec)
implicit none
class(cFileArr)      :: this

real, intent(IN)     :: rtime
integer, intent(OUT) :: year,month,day,hour,minute,isec,msec

type(datetime)       :: pcas
integer              :: rtime_m

    rtime_m = nint(rtime*1000)
    write(*,*) rtime_m
    pcas = this%reftime + timedelta(0,0,0,0,rtime_m)
    write(*,*) pcas%isoformat()
    year =  pcas%getYear()
    month = pcas%getMonth()
    day   = pcas%getDay()
    hour  = pcas%getHour()
    minute= pcas%getMinute()
    isec  = pcas%getSecond()
    msec  = pcas%getMillisecond()

end subroutine cfilearr_todatetime

!
subroutine cfilearr_fill(this, hyp)
!
use stations,   only: nstat, stat_name, xstat, ystat, zstat
!
   implicit none

   class(cFileArr)      :: this
   type (cFileHyp)      :: hyp

!
   integer i
   integer j
!
   real    pwt
   real    psum, psum2, mwt
   type(datetime)             :: pcas, mcas
   type(timedelta)            :: dcas
   type(cRecordHyp)           :: r
   type(cRecordArr)           :: a
! reftime
   mcas = datetime(3000)
   do i=1,hyp%n_rec
     r = hyp%rec(i)
     pcas = datetime(r%getYear(),r%month,r%day,&
        &r%hour,r%minute,r%isec,r%msec)
     if (pcas .lt. mcas) mcas=pcas
   end do
!   if(mcas%getSecond() .lt. 30) then
!      mcas = mcas - timedelta(0,0,1)
!   endif
   this%reftime = datetime(mcas%getYear(),mcas%getMonth(),mcas%getDay(),mcas%getHour(),mcas%getMinute())
   write(*,*) this%reftime%isoformat()

if (allocated(this%arr)) deallocate(this%arr)
allocate(this%arr(hyp%n_rec))

   this%n_arr=0
   do i=1,hyp%n_rec
     r = hyp%rec(i)
!     if (r%wt .eq. 4) cycle
     a%id = r%id
     a%sta = r%sta
     a%phase = r%phase
     a%amp = r%amp_v
     if (r%period > 0) then
       a%freq = 1.0/r%period
     else
       a%freq = 99.9
     end if

! compute weights
     pwt = r%wt
! weights refer (are inverse proportional) to the standard deviation
     if (hyp%hyr) then
        if (pwt .gt. 0) then
             pwt = pwt/1000.0
             pwt = 1. / pwt
        else if (pwt .lt. 0) then
             pwt = 0.0
        else
             pwt = 0.0
        endif
    else
             pwt=(4.-pwt)/4.
    endif
    a%wt = pwt

     pcas = datetime(r%getYear(),r%month,r%day,&
        &r%hour,r%minute,r%isec,r%msec)
     dcas = pcas - this%reftime
     a%trec = real(dcas%total_seconds())
! key
     a%key=0
     do j=1,nstat
         if (a%sta .eq. stat_name(j)) then
             a%key=j
             a%X=xstat(j)
             a%Y=ystat(j)
             a%Z=zstat(j)
         endif
     end do
     if (a%key .eq. 0) then
         write (*,'(1x,"Station ",a5," not found.")')&
               &   this%arr(i)%sta
         cycle
     end if
     this%arr(i)=a
     this%n_arr=this%n_arr+1
   end do   !i n_rec

!
!   sum weights
!
   psum=0.0
   psum2=0.0
   mwt = 0.0
   do  i=1,this%n_arr
      pwt=this%arr(i)%wt
      psum=psum+pwt
      psum2=psum2+pwt*pwt
      if (pwt > mwt) mwt = pwt
   end do  !i n_arr
   this%sumwt=psum
   this%sumwt2=psum2
   this%maxwt=mwt
!
return

end subroutine cfilearr_fill

!      if (type(i) .eq. 'S') then
!      else if (type(i) .eq. 'P') then
!      else
!         write (*,'(1x," : Wrong type of arrival in record ",i2)') i
!      endif
end module Arrivals

File: libsrc/arrivals.cpp
#include <iostream>
#include <vector>
#include <string>
#include <cmath>
#include <cstdlib>
#include <ctime>
#include <chrono>
#include <iomanip>
#include <sstream>
#include <map>

#include "hypfile.h"
//#include "datetime.hpp"
#include "stations.h"
/*
struct TArrival {
    int id;
    std::string sta;
    std::string phase;
    double trec;
    double wt;
    double amp;
    double freq;
    double X;
    double Y;
    double Z;
    double delayP;
    double delayS;
};

class CArrivalFile {
public:
    int n_arr = 0;
    double reftime;
    double sumwt = 0.0;
    double sumwt2 = 0.0;
    double maxwt = 0.0;
    std::vector<TArrival> arr;
    
    void fill(const std::vector<TRecordHyp> &hyp, const std::map<std::string,TStation> &stations);
};
*/

#include "arrivals.h"


// function for converting time and date to epoch time
// input integer numbers of: year, month, day, hour, minute, second, millisecond
// output double number of seconds since epoch
// epoch time is the number of seconds since 1970-01-01 00:00:00 UTC
// both input and output are in UTC
double toEpochTime(int year, int month, int day, int hour, int minute, int second, int millisecond) {
    std::tm tm = {};
    tm.tm_year = year - 1900; // tm_year is years since 1900
    tm.tm_mon = month - 1;    // tm_mon is months since January (0-11)
    tm.tm_mday = day;
    tm.tm_hour = hour;
    tm.tm_min = minute;
    tm.tm_sec = second;
    tm.tm_isdst=-1;
    tm.tm_zone = NULL;
    tm.tm_gmtoff = 0;
    
    // Convert to time_t (seconds since epoch)
    //std::time_t time_since_epoch = std::mktime(&tm);
    //std::time_t time_since_epoch = std::timelocal(&tm);
    std::time_t time_since_epoch = timelocal(&tm);
    
    // Add milliseconds
    double epoch_time = static_cast<double>(time_since_epoch) + static_cast<double>(millisecond) / 1000.0;
    
    return epoch_time;
}


CArrivalFile arrs;
CArrivalFile* parrs = &arrs;

void CArrivalFile::fill(const std::vector<TRecordHyp> &hyp, const std::map<std::string,TStation> &stations){

// fill std::vector<TArrival> arr with records computed from all records in hyp joinded with stations on station name
// hyp is a vector of TRecordHyp
// stations is a map of station name to TStation
// arr is a vector of TArrival
// fill only arr.sta, arr.phase, arr.freq, arr.X, arr.Y, arr.Z, arr.delayP, arr.delayS
// arr.freq is computed from hyp.period
// arr.X, arr.Y, arr.Z, arr.delayP, arr.delayS are taken from stations
// arr.wt is computed from hyp.wt

// test hyp.wt content and set flag hyr
// if one of hyp.wt are not in enumerated set of integer numbers {0,1,2,3,4}
// then set hyr flag to true else set it to false
   bool hyr = false; // initialize hyr to false
   for (const auto& r : hyp) {
      if (r.wt < 0 || r.wt > 4) {
	  // If any weight is outside the range [0, 4], set hyr to true
	  hyr = true;
	  break;
      }
      // check if r.wt is an integer
      if (r.wt != static_cast<int>(r.wt)) {
          // if r.wt is not integer number then set hyr to true
          hyr = true;
          break;
      }
   }

   n_arr = 0; // reset number of arrivals
	     
   for (const auto& r : hyp) {
	TArrival a;
	a.id = n_arr+1; // set id to current number of arrivals + 1
	a.sta = r.sta;
	a.phase = r.phase;
	a.trec = 0.0; // initialize to zero
	a.wt = 0.0; // initialize to zero
	a.amp = r.amp_v;

        // convert r.year from YY format to YYYY format
	// r.year is in YY format
	// int yyyy to the year in YYYY format
	int yyyy;
        if (r.year < 70) {
	    yyyy = r.year + 2000; // convert to YYYY format
	} else {
	    yyyy = r.year + 1900; // convert to YYYY format
	}
	// compute epoch time a.trec from fields of TRecordHyp
 
	a.trec = toEpochTime(yyyy, r.month, r.day, r.hour, r.minute, r.isec, r.msec);

	// find station in stations map
	auto it = stations.find(r.sta);
	if (it != stations.end()) {
	    const TStation& station = it->second;
	    a.X = station.Xk;
	    a.Y = station.Yk;
	    a.Z = station.elevation;
	    a.delayP = station.delayP;
	    a.delayS = station.delayS;
	} else {
	    std::cerr << "Station " << r.sta << " not found in stations map." << std::endl;
	    continue; // skip this record if station not found
	}
	
	// compute frequency
	if (r.period > 0) {
	    a.freq = 1.0 / r.period;
	} else {
	    a.freq = 99.9; // default value for frequency
	}
	
	// compute weights
	double pwt = r.wt;
	if (hyr) {
	// hyp.wt is in miliseconds
	    if (pwt > 0) {
		pwt = pwt / 1000.0;
		pwt = 1.0 / pwt;
	    } else if (pwt < 0) {
		pwt = 0.0; // negative weight set to zero
	    } else {
		pwt = 0.0; // zero weight remains zero
	    }
	} else {
	// hyp.wt is acording HYPO71
	    pwt = (4.0 - pwt) / 4.0; // normalize weight
	}
	
	a.wt = pwt; // assign computed weight to arrival
	
	arr.push_back(a); // add arrival to vector
	n_arr++; // increment number of arrivals
   } // end of loop over hyp records
   // now fill the other members of the class
   // reftime is the minimal time of the first arrival rounded to the floor to whole tents
   reftime = arr[0].trec; // initialize
   for (int i = 0; i < n_arr; i++) {
      if (arr[i].trec < reftime) {
      reftime = arr[i].trec;
      }
   }
   // reftime round to the floor to whole tens
   reftime = floor(reftime/10.0)*10.0; // round to the floor of the second

   // sum weights
   sumwt = 0.0;
   sumwt2 = 0.0;
   maxwt = 0.0;
   for(int i = 0; i < n_arr; i++) {
	// arr[i].trec = arr[i].trec - reftime; // remove reference time
      double pwt = arr[i].wt;
      sumwt = sumwt + pwt;
      sumwt2 = sumwt2 + pwt * pwt;
      if(pwt > maxwt) maxwt = pwt;
    }

   for(int i = 0; i < n_arr; i++) {
      arr[i].trec = arr[i].trec - reftime; // remove reference time
     // arr[i].wt = arr[i].wt / sumwt; // normalize weights
   }
    // std::cout << "Number of arrivals: " << n_arr << " / "  << arr.size() << std::endl;
    // std::cout << "reftime = " << reftime << std::endl;
    // std::cout << "sumwt = " << sumwt << std::endl;
    // std::cout << "sumwt2 = " << sumwt2 << std::endl;
    // std::cout << "maxwt = " << maxwt << std::endl;
}

/*    
    // Demonstrate toDateTime conversion for a sample rtime (e.g., 1.234 seconds)
    int year, month, day, hour, minute, isec, msec;
    arrs.toDateTime(1.234, year, month, day, hour, minute, isec, msec);
    
    // Output the computed weight sums
    std::cout << "n_arr = " << arrs.n_arr << std::endl;
    std::cout << "sumwt = " << arrs.sumwt << std::endl;
    std::cout << "sumwt2 = " << arrs.sumwt2 << std::endl;
    std::cout << "maxwt = " << arrs.maxwt << std::endl;
    
    return 0;
*/
